mlos_bench/mlos_bench/services/config_persistence.py

```diff
@@ -120,7 +126,19 @@ def load_config(self, json_file_name: str) -> Union[Dict[str, Any], List[Dict[st
         json_file_name = self.resolve_path(json_file_name)
         _LOG.info("Load config: %s", json_file_name)
         with open(json_file_name, mode='r', encoding='utf-8') as fh_json:
-            return json5.load(fh_json)  # type: ignore[no-any-return]
+            config = json5.load(fh_json)
+        if schema_type is not None:
+            try:
+                schema_type.validate(config)
+            except Exception as ex:
+                _LOG.error("Failed to validate config %s against schema type %s at %s",
+                           json_file_name, schema_type.name, schema_type.value)
+                raise ex
+        if isinstance(config, dict) and config.get("$schema"):
+            # Remove $schema attributes from the config after we've validated
+            # them to avoid passing them on to other objects.
+            del config["$schema"]
```

Comment:
Yeah, I've thought about that.  I see a few options atm:

1. Make them objects that include a list of environments instead of just an array.  This would be more similar to the way that `-tunables.jsonc` files work for tunable params now.  For instance:
```json
{
    "$schema": "mlos_bench/schemas/environments-schema.json", // optional
    "description": "a list of environments for X",
    "environments": {  // could be either a dict or a simple list (I don't think we'd use the "name" part other than for descriptive purposes)
        "env-1": {
            // either a full config or maybe even a reference to another includeable file?
            ...
        },
        "env-2": {
            ...
        },
    },
}
```

2. Just omit the `$schema` attribute in those files (and potentially lose some editor abilities for configs in external repos).
3. For all environment configs to be individual files and then compose the related ones together as a composite env somehow that references them and then have nested composite envs (I haven't worked through this idea fully yet, but have wonderered if it might show up again for multiple VMs later as well).

My strategy for now was to go with option 2 initially, and then have a separate PR to convert to option 1 if you agree.  Thoughts?

Reply:
Awesome!

I haven't written the schemas for Environments yet, so we can finish discussing this in that PR next.  I was trying to scope it down to smaller portions of the code/configs at a time.